bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java

```diff
@@ -278,7 +278,7 @@ private void decodeProperties(final BsonReader reader, final DecoderContext deco
         boolean discriminatorEnabled = changeTheDiscriminator ? propertyModel.useDiscriminator() : clazzModel.useDiscriminator();
         return new ClassModel<S>(clazzModel.getType(), clazzModel.getPropertyNameToTypeParameterMap(),
                 clazzModel.getInstanceCreatorFactory(), discriminatorEnabled, clazzModel.getDiscriminatorKey(),
-                clazzModel.getDiscriminator(), concreteIdProperty, concretePropertyModels);
+                clazzModel.getDiscriminator(), IdPropertyModelHolder.create(clazzModel, concreteIdProperty), concretePropertyModels);
```

Comment:
Maybe add a test of id generation with a class that has a type parameter to test this path?

Reply:
This isn't necessary, the type parameter would have to be concrete and that wouldn't test this path. This path triggers when there is a change to a classmodel due to it being used as a property inside another Pojo. For example setting the discriminator.